### PR TITLE
New Python plotting script for creating difference plots

### PR DESCRIPTION
**Pull request by BenjaminBlake-NOAA**
# DESCRIPTION OF CHANGES: 
This pull request adds a new Python plotting script for creating difference plots with cartopy using two grib2 files.  Also, I'm not sure why the QPF plots were commented out in plot_allvars.py because they work so I've uncommented them.

## TESTS CONDUCTED: 
This script was tested on Hera.

## ISSUE (optional): 
#263 

## CONTRIBUTORS (optional): 
David Wright also contributed to the original plot_allvars.py script.

### CODE REVIEW

**Comment by dmwright526** [comment]
@BenjaminBlake-NOAA This looks great! I kept commenting out the QPF because it isn't in the output I have been testing with, but that appears to be an issue with my output...

I think it would be good idea to test this on the 25km CONUS grid output (if you haven't already), specifically to make sure there is a high enough density of wind barbs on the 250 and 500mb plots.

**Comment by BenjaminBlake-NOAA** [comment]
@dmwright526 Ok that makes sense about the QPF code!  I have not tested it with 25-km output yet.

**Comment by JeffBeck-NOAA** [comment]
Hey @BenjaminBlake-NOAA @dmwright526, I'm going to test this script with my 25-km output right now.

**Comment by JeffBeck-NOAA** [comment]
@BenjaminBlake-NOAA @dmwright526 

I see APCP in my grib2 files, so I don't know if this error is related to QPF plotting, but I'm getting this error when running the difference Python script on Hera:

Working on composite reflectivity for conus
multiprocessing.pool.RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/contrib/anaconda/anaconda3/latest/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/contrib/anaconda/anaconda3/latest/lib/python3.7/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "plot_allvars_dif.py", line 1058, in plot_all
    csdif = ax3.contourf(lon_shift,lat_shift,refc_1,clevsdif,colors='red',transform=transform)
  File "/contrib/anaconda/anaconda3/latest/lib/python3.7/site-packages/cartopy/mpl/geoaxes.py", line 1404, in contourf
    for col in result.collections
  File "/contrib/anaconda/anaconda3/latest/lib/python3.7/site-packages/matplotlib/transforms.py", line 701, in union
    raise ValueError("'bboxes' cannot be empty")
ValueError: 'bboxes' cannot be empty
"""

Can we change all "dif" to "diff" in the script and within the script name?  I'm also not seeing wind barbs in the 250 mb wind plots (plus a problem with the dam contours in all the plots):

Screen Shot 2020-08-28 at 11 37 07 AM

File: ush/Python/plot_allvars_dif.py [hunk 1]
```diff
@@ -0,0 +1,1077 @@
...
+EXPT_BASEDIR_1 = str(sys.argv[3])
+EXPT_SUBDIR_1 = str(sys.argv[4])
+EXPT_BASEDIR_2 = str(sys.argv[5])
+EXPT_SUBDIR_2 = str(sys.argv[6])
+CARTOPY_DIR = str(sys.argv[7])
+
+# Define the location of the input files
+data1 = pygrib.open(EXPT_BASEDIR_1+'/'+EXPT_SUBDIR_1+'/'+ymdh+'/postprd/RRFS.t'+cyc+'z.bgdawp'+fhour+'.tm00.grib2')
+data2 = pygrib.open(EXPT_BASEDIR_2+'/'+EXPT_SUBDIR_2+'/'+ymdh+'/postprd/RRFS.t'+cyc+'z.bgdawp'+fhour+'.tm00.grib2')
+
```
Comment (JeffBeck-NOAA) [inline ush/Python/plot_allvars_dif.py]: We need to change "RRFS" to "rrfs" and "bgdawp" to "bgdawpf" to match the new post_fix PR that I'm about to issue.  Thanks!

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA  Yes we can change dif to diff and make those changes to the filenames, that's fine with me!  The error you are showing appears to be related to the reflectivity plot.  Since it doesn't really make sense to have a difference plot for reflectivity I made the third panel show color fill for both models.  Here's an example from my website.  I can also look into the wind barb issue when I get a chance, I don't recall if I had the same issue with my tests.
comparerefc_conus_f14

**Comment by dmwright526** [comment]
@BenjaminBlake-NOAA I think the wind barb issue at 250mb and 500mb has to do with the "skip" variable used for thinning the data. If I recall correctly, it is set at 70, so when using the 25km grid it may only show 1-2 wind barbs. We can try copying over what we had done for the 10m winds to see if it is thinned out enough.

**Comment by JeffBeck-NOAA** [comment]
@BenjaminBlake-NOAA @dmwright526 I was able to fix the wind barb plotting problem by using David's "skip" code from the 10-m winds and adding it to the 500- and 250-mb plots.  We can probably just add that code to this PR for the original plotting script, in addition to adding it to this different script.  @BenjaminBlake-NOAA, your reflectivity color fill replacement for a "diff" plot makes sense and is a great idea.  I'm curious to see if I get the same error when processing the reflectivity with a 3-km domain, or if it's another issue (in addition to the contour plotting) of having NaNs in the write component domain with the 25-km domain.

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA  Glad you were able to resolve the issue with the wind barbs, and adding the skip code to the original plotting script sounds good to me.  I wouldn't be surprised if NaNs are also causing an issue with the reflectivity color fill plot.  I did not get an error when plotting with my 3-km data so I'm curious to see what you find there.

**Comment by jwolff-ncar** [comment]
Related to issue #263 (https://github.com/NOAA-EMC/regional_workflow/issues/263)

**Comment by JeffBeck-NOAA** [comment]
Hey @BenjaminBlake-NOAA, @dmwright526, I was able to get around to plotting the three pre-defined domains on Jet.  Please see my comments below.  Hopefully these things aren't hard to implement/fix.

- The 3-km domain works great for both the original and difference plots, and all wind barbs/contours look reasonable.
- Just an FYI: I have a version of the CARTOPY files now on Jet here: /lfs4/BMC/wrfruc/beck/NaturalEarth
- For all comments below, please refer to plots from both scripts for all three pre-defined domains on Jet here:
/lfs4/BMC/wrfruc/beck/Python/ush/Python/3-km
/lfs4/BMC/wrfruc/beck/Python/ush/Python/13-km
/lfs4/BMC/wrfruc/beck/Python/ush/Python/25-km

**In plot_allvars.py:** 

1. For the post filenames, can you change "RRFS" to "rrfs", change the hour from two to three digits (also needs to be updated in the instructions and command line variable as three digits), and add an "f" before the hour?  These look good in the diff script, but I guess the original Python script never got updated.
2. The winds for the 500-mb and 250-mb plots are too sparsely displayed at 13-km and 25-km.

**In plot_allvars_diff.py:** 

1. Change the REFC plot names from "dif" to "diff".
2. The winds for the upper-level plots look good at 13-km and 25-km, but the decameter contours are missing in the difference plot.
3. Some good news, I'm no longer getting the errors at 25-km that I was previously.

**For both scripts:**

1. Do you have any objection about combining the EXPT_BASEDIR and EXPT_SUBDIR into a single command line argument (say, EXPT_DIR?).  I find that I copy the full path to the experiment and then have to add a space to make sure I have the two arguments satisfied.  It seems like it might be easier to just make it a single argument?  Or maybe there was a reason you defined it the way it is?

2. Note that on (at least) Jet, the user will need to run a "module purge" command and then the following to run the Python scripts:

```
module use /contrib/miniconda3/modulefiles
module load miniconda3
conda activate adb_graphics
```

Since two different conda "modules" are required to run the workflow vs the Python scripts (adb_graphics vs regional_workflow), we probably want to mention this in the documentation.  Otherwise, users will get the "pygrib" error when trying to run the Python plotting scripts after they've run the workflow.

3. Is there a way to make the default usage output a bit clearer when you type the name of the script in the command line?  Is it possible to make this a (bulleted?) list?  Right now it looks like this (for the Python difference script): 

usage: plot_allvars.py [-h]
Cycle date/time in YYYYMMDDHH format Forecast hour in HHH format Path to
                            experiment 1 base directory Path to experiment 1 named directory Path to
                            experiment 2 base directory Path to experiment 2 named directory Path to
                            base directory of cartopy shapefiles
plot_allvars_diff.py: error: the following arguments are required: Cycle date/time in YYYYMMDDHH format, Forecast hour in HHH format, Path to experiment 1 base directory, Path to experiment 1 named directory, Path to experiment 2 base directory, Path to experiment 2 named directory, Path to base directory of cartopy shapefiles

**Comment by dmwright526** [comment]
@JeffBeck-NOAA glad to hear that it is running! @BenjaminBlake-NOAA I can take a look at things this morning.

I agree there will need to be some additional documentation for the different platforms to get the right environment loaded. I know regional_workflow doesn't currently have the right packages installed with it, but I am not sure what environments we have access to on platforms outside of Hera and Jet.

**Comment by BenjaminBlake-NOAA** [comment]
Hi @JeffBeck-NOAA and @dmwright526 , I can go ahead and make the following changes this morning.  They should be easy to implement and I agree with all of them!

In plot_allvars.py:

For the post filenames, can you change "RRFS" to "rrfs", change the hour from two to three digits (also needs to be updated in the instructions and command line variable as three digits), and add an "f" before the hour? These look good in the diff script, but I guess the original Python script never got updated.

In plot_allvars_diff.py:

Change the REFC plot names from "dif" to "diff".
The winds for the upper-level plots look good at 13-km and 25-km, but the decameter contours are missing in the difference plot.

For both scripts:

Do you have any objection about combining the EXPT_BASEDIR and EXPT_SUBDIR into a single command line argument (say, EXPT_DIR?). I find that I copy the full path to the experiment and then have to add a space to make sure I have the two arguments satisfied. It seems like it might be easier to just make it a single argument? Or maybe there was a reason you defined it the way it is?

David, would you be able to look into the winds for the 500-mb and 250-mb plots being too sparsely displayed at 13-km and 25-km for the plot_allvars.py script?

**Comment by dmwright526** [comment]
@BenjaminBlake-NOAA sure thing.

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA @dmwright526  I just added the following changes to the scripts:

In plot_allvars.py: Update post filenames and change forecast hour from two to three digits

In plot_allvars_diff.py: Change the REFC plot names from "dif" to "diff".

For both scripts: Combining EXPT_BASEDIR and EXPT_SUBDIR into EXPT_DIR

I forgot to look into why the 500 decameter contours are missing from the difference plot, I'll do that shortly.

**Comment by BenjaminBlake-NOAA** [comment]
500_diff_conus_f024

@JeffBeck-NOAA In my 500-mb difference plot, I have contour values for the height differences (color fill).  Are you referring to the contour lines not being there?  If so, then you are correct because I never added them.  :)  Is that something we want to include in addition to the color fill differences?  Note that for this example I used two forecasts from different days so the pattern is very different!

**Comment by JeffBeck-NOAA** [comment]
@dmwright526, @BenjaminBlake-NOAA, thanks for your work on the PR this morning!  I will retest my three cases, and for the difference plots, I'll run a completely different day.  Now that I know there aren't any decameter contours for the difference plots, I assume that using the same day between my two cases and only varying the physics suites isn't sufficient to show any differences!  @BenjaminBlake-NOAA, one quick request: In the difference plots, do you mind changing FV3-LAM-X to FV3-LAM-2?  Since this is just a second version and not necessarily the "experimental" FV3-LAM, I think that change makes sense?

**Comment by JeffBeck-NOAA** [comment]
> 500_diff_conus_f024
> 
> @JeffBeck-NOAA In my 500-mb difference plot, I have contour values for the height differences (color fill). Are you referring to the contour lines not being there? If so, then you are correct because I never added them. :) Is that something we want to include in addition to the color fill differences? Note that for this example I used two forecasts from different days so the pattern is very different!

Hey @BenjaminBlake-NOAA, @dmwright526, do you guys know of a way to scale the limits of the colorbar based on the extremes of the contour values?  Just wondering if this can be done on the fly based on the data, so that the plot isn't washed out.  Thanks!

**Comment by JeffBeck-NOAA** [comment]
@BenjaminBlake-NOAA, how did you run the difference script from two different dates when there's only a single YYYYMMDDHH command line argument?  Shouldn't we modify that script to require two sets of dates and forecast hours?  So it'd be something like:

python plot_allvars_diff.py 2019070100 006 /lfs4/BMC/wrfruc/beck/expt_dirs/test_RRFS_CONUS_3km 2019080100 006 /lfs4/BMC/wrfruc/beck/expt_dirs/test_RRFS_CONUS_3km_RRFS_v1beta /lfs4/BMC/wrfruc/beck/NaturalEarth

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA FV3-LAM-2 is fine by me, I can make that change!

There should be a way to scale the limits of the colorbar based on the max/min values.  I imagine there are certain plots where this would be desirable.  I can try and work on that when I get a chance unless someone else gets to it before me.

For running the difference script with two different dates, I tricked the script into thinking it was for the same date by putting the file into a directory with the same date but different experiment name.  That was just so I could see differences.  :)  Normally I'd think we would want to compare the same date / forecast hour though, right?  We could specify it twice though in case we wanted to compare different dates / forecast hours

**Comment by JeffBeck-NOAA** [comment]
> @JeffBeck-NOAA FV3-LAM-2 is fine by me, I can make that change!
> 
> There should be a way to scale the limits of the colorbar based on the max/min values. I imagine there are certain plots where this would be desirable. I can try and work on that when I get a chance unless someone else gets to it before me.
> 
> For running the difference script with two different dates, I tricked the script into thinking it was for the same date by putting the file into a directory with the same date but different experiment name. That was just so I could see differences. :) Normally I'd think we would want to compare the same date / forecast hour though, right? We could specify it twice though in case we wanted to compare different dates / forecast hours

Thanks, Ben!  Yeah, normally we'd want to compare the same day, but to make it clear, I think adding the option to define the date/forecast hour for both directories makes sense.  

Scaling the colorbar limits is desirable, but probably not absolutely necessary for the release.  I thought maybe you or @dmwright526 might have some pre-existing Python code that implements this (I do for NCL, but that won't help), so if not, I think we can just leave it as is for now.

**Comment by dmwright526** [comment]
@JeffBeck-NOAA I don't have any code lying around for this. The difference script is mainly being developed for use with the Graduate Student Test, is it not? Do we know what that test will look like exactly and if this script, as it stands now, will be able to display those changes?

**Comment by JeffBeck-NOAA** [comment]
> @JeffBeck-NOAA I don't have any code lying around for this. The difference script is mainly being developed for use with the Graduate Student Test, is it not? Do we know what that test will look like exactly and if this script, as it stands now, will be able to display those changes?

Thanks, @dmwright526.  Yes, this script was a request by @jwolff-ncar for the GST.  @jwolff-ncar , are there specific requirements that must be met regarding the difference script?  Right now each variable plot consists of displaying two subplots (from each of the two runs) and then a difference plot with contour values.  We are expanding the command line arguments to account for different dates/forecast hours, as previously users were only able to plot differences for the same date/forecast hour.

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA  I added the changes.  However, since there are two specified dates I wasn't sure how to handle that for the label on the difference panel.  Maybe we leave the date off of that panel entirely?  For now I left it as the date of the first file

**Comment by JeffBeck-NOAA** [comment]
> @JeffBeck-NOAA I added the changes. However, since there are two specified dates I wasn't sure how to handle that for the label on the difference panel. Maybe we leave the date off of that panel entirely? For now I left it as the date of the first file

@BenjaminBlake-NOAA , Thanks!  What do you think of:

Initialized: YYYYHHMMDD (FV3-LAM), YYYYHHMMDD (FV3-LAM-2), Valid: FHH (FV3-LAM), FHH (FV3-LAM-2)

**Comment by JeffBeck-NOAA** [comment]
@BenjaminBlake-NOAA , @dmwright526 , the good news is that all plots look reasonable for all three pre-defined domains!  I think we can safely merge this PR after we finalize the difference plot headers.

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA Great to hear about all the plots looking reasonable!  Yeah that sounds good for the difference plot headers, I will make that change

**Comment by JeffBeck-NOAA** [comment]
> @JeffBeck-NOAA Great to hear about all the plots looking reasonable! Yeah that sounds good for the difference plot headers, I will make that change

Thanks, Ben!  I'll pull that change, run a quick test, then let you merge this PR!

**Comment by BenjaminBlake-NOAA** [comment]
@JeffBeck-NOAA  Ok I added the change, let me know after you run your test and I will merge the PR!

**Review by JeffBeck-NOAA — APPROVED** [review]

**Comment by JeffBeck-NOAA** [comment]
@BenjaminBlake-NOAA, the new plot looks good!  Feel free to merge.

**Pull request merged**